crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -285,150 +205,197 @@ impl IndexAllocator {
             rand::thread_rng().gen()
         };
         let rng = SmallRng::from_seed(seed);
-        IndexAllocator(Mutex::new(Inner { rng, state }))
+        IndexAllocator(Mutex::new(Inner {
+            rng,
+            strategy,
+            free_list: ids,
+            per_module: HashMap::new(),
+            slot_state: (0..max_instances)
+                .map(|i| {
+                    SlotState::Free(FreeSlotState::NoAffinity {
+                        free_list_index: GlobalFreeListIndex(i),
+                    })
+                })
+                .collect(),
+        }))
     }
 
-    /// Allocate a new slot.
+    /// Allocate a new index from this allocator optionally using `id` as an
+    /// affinity request if the allocation strategy supports it.
+    ///
+    /// Returns `None` if no more slots are available.
     pub fn alloc(&self, id: Option<CompiledModuleId>) -> Option<SlotId> {
+        self._alloc(id, false)
+    }
+
+    /// Attempts to allocate a guaranteed-affine slot to the module `id`
+    /// specified.
+    ///
+    /// Returns `None` if there are no slots affine to `id`. The allocation of
+    /// this slot will not record the affinity to `id`, instead simply listing
+    /// it as taken. This is intended to be used for clearing out all affine
+    /// slots to a module.
+    pub fn alloc_affine_and_clear_affinity(&self, id: CompiledModuleId) -> Option<SlotId> {
+        self._alloc(Some(id), true)
+    }
+
+    fn _alloc(
+        &self,
+        id: Option<CompiledModuleId>,
+        force_affine_and_clear_affinity: bool,
```

Comment:
Can we make this a little indicator enum, e.g. `enum AllocMode { ForceAffineAndClear, AnySlot }` or somesuch?